subsys/net/lib/nrf_provisioning/src/nrf_provisioning_coap.c

```diff
@@ -0,0 +1,643 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <modem/modem_info.h>
+#include <modem/nrf_modem_lib.h>
+
+#include <zephyr/net/socket.h>
+#include <zephyr/net/tls_credentials.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/net/coap_client.h>
+
+#include "nrf_provisioning_at.h"
+#include "nrf_provisioning_codec.h"
+#include "nrf_provisioning_coap.h"
+#include "nrf_provisioning_jwt.h"
+
+LOG_MODULE_REGISTER(nrf_provisioning_coap, CONFIG_NRF_PROVISIONING_LOG_LEVEL);
+
+#if !defined(CONFIG_NRF_PROVISIONING_ROOT_CA_SEC_TAG)
+#error "CONFIG_NRF_PROVISIONING_ROOT_CA_SEC_TAG not given"
+#endif
+
+#define TLS_SEC_TAG CONFIG_NRF_PROVISIONING_ROOT_CA_SEC_TAG
+#define COAP_HOST CONFIG_NRF_PROVISIONING_COAP_HOSTNAME
+#define PEER_PORT CONFIG_NRF_PROVISIONING_COAP_PORT
+
+#define MODEM_INFO_FWVER_SIZE 41
+#define CLIENT_VERSION "1"
+
+#define AUTH_MVER "mver=%s"
+#define AUTH_CVER "cver=%s"
+#define AUTH_PATH "p/auth"
+#define AUTH_API_TEMPLATE (AUTH_PATH "?" AUTH_MVER "&" AUTH_CVER)
+
+#define CMDS_PATH "p/cmd"
+#define CMDS_AFTER "after=%s"
+#define CMDS_MAX_RX_SZ "rxMaxSize=%s"
+#define CMDS_MAX_TX_SZ "txMaxSize=%s"
+#define CMDS_API_TEMPLATE (CMDS_PATH "?" CMDS_AFTER "&" CMDS_MAX_RX_SZ "&" CMDS_MAX_TX_SZ)
+
+static const char *resp_path = "p/rsp";
+
+static struct addrinfo *address;
+static struct coap_client client;
+
+static K_SEM_DEFINE(coap_response, 0, 1);
+
+int nrf_provisioning_coap_init(struct nrf_provisioning_coap_mm_change *mmode)
+{
+	static bool initialized;
+	int ret;
+
+	nrf_provisioning_codec_init((struct nrf_provisioning_codec_mm_change *)mmode);
+
+	if (initialized) {
+		return 0;
+	}
+
+	if (IS_ENABLED(CONFIG_NRF_PROVISIONING_SYS_INIT)) {
+		ret = nrf_modem_lib_init();
+		if (ret < 0) {
+			LOG_ERR("Failed to initialize modem library, error: %d", ret);
+			return ret;
+		}
+	}
+
+	/* Modem info library is used to obtain the modem FW version */
+	ret = modem_info_init();
+	if (ret) {
+		LOG_ERR("Modem info initialization failed, error: %d", ret);
+		return ret;
+	}
+
+	LOG_INF("Init CoAP client");
+	ret = coap_client_init(&client, NULL);
+	if (ret) {
+		LOG_ERR("Failed to initialize CoAP client, error: %d", ret);
+		return ret;
+	}
+
+	initialized = true;
+
+	return 0;
+}
+
+/* Setup TLS options on a given socket */
+static int tls_setup(int fd)
```

Comment:
If you intend to support either `SOCK_DGRAM` or `SOCK_STREAM` (UDP or TCP), then `tls_setup()` is fine.  But if this is really only intended for `SOCK_DGRAM`, then `dtls_setup()` seems more appropriate.